src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java

```diff
@@ -159,6 +169,41 @@ void rollback(DeploymentDocument deploymentDocument, CompletableFuture<Deploymen
         }
     }
 
+    /**
+     * Finds all services which are dependers of given broken services, directly or indirectly
+     * This method performs a breadth-first search, starting from the broken services and traversing through
+     * service dependencies.
+     * @param rollbackServices the set of rollback services to track
+     * @param brokenServiceNames the set of broken service names
+     * @return a set of all services depending on the broken services, including themselves
+     */
+    private Set<GreengrassService> findServiceDependers(final Set<GreengrassService> rollbackServices,
+                                                        final Set<String> brokenServiceNames) {
+
+        Set<GreengrassService> dependerServices = rollbackServices.stream()
+                .filter(service -> brokenServiceNames.contains(service.getName()))
+                .collect(Collectors.toSet());
+        Queue<GreengrassService> dependers = new LinkedList<>(dependerServices);
+
+        // Breadth-first search to find all dependent services, staring from broken services
+        while (!dependers.isEmpty()) {
+            GreengrassService currentService = dependers.poll();
+            for (GreengrassService depender : currentService.getDependers(false)) {
+                // Ensure dependers are within the rollback services and haven't been processed
+                if (rollbackServices.contains(depender) && dependerServices.add(depender)
```

Comment:
I don't think this `if` is doing anything helpful. The queue has no max size, so `offer` cannot fail.

Would the straightline code work just as well?

```
dependerServices.add(depender);
dependers.offer(depender);
```

Reply:
Updated